cpp/include/cudf/table/experimental/row_operators.cuh

```diff
@@ -0,0 +1,421 @@
+#include <cudf/sorting.hpp>
+#include <cudf/table/row_operators.cuh>
+#include <cudf/table/table_device_view.cuh>
+#include <cudf/utilities/traits.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+
+#include <thrust/equal.h>
+#include <thrust/swap.h>
+#include <thrust/transform_reduce.h>
+
+#include <cuda/std/tuple>
+#include <cuda/std/utility>
+
+#include <limits>
+#include <memory>
+#include <utility>
+
+namespace cudf {
+namespace experimental {
+
+/**
+ * @brief A map from cudf::type_id to cudf type that excludes LIST and STRUCT types.
+ *
+ * To be used with type_dispatcher in place of the default map, when it is required that STRUCT and
+ * LIST map to void. This is useful when we want to avoid recursion in a functor. For example, in
+ * element_comparator, we have a specialization for STRUCT but the type_dispatcher in it is only
+ * used to dispatch to the same functor for non-nested types. Even when we're guaranteed to not have
+ * non-nested types at that point, the compiler doesn't know this and would try to create recursive
+ * code which is very slow.
+ *
+ * Usage:
+ * @code
+ * type_dispatcher<dispatch_nested_to_void>(data_type(), functor{});
+ * @endcode
+ */
+template <cudf::type_id t>
+struct dispatch_nested_to_void {
+  using type = std::conditional_t<cudf::is_nested(data_type(t)), void, id_to_type<t>>;
+};
+
+namespace lex {
+
+/**
+ * @brief Performs a relational comparison between two elements in two columns.
+ *
+ * @tparam Nullate A cudf::nullate type describing how to check for nulls.
+ */
+template <typename Nullate>
+class element_comparator {
+ public:
+  /**
+   * @brief Construct type-dispatched function object for performing a
+   * relational comparison between two elements.
+   *
+   * @note `lhs` and `rhs` may be the same.
+   *
+   * @param has_nulls Indicates if either input column contains nulls.
+   * @param lhs The column containing the first element
+   * @param rhs The column containing the second element (may be the same as lhs)
+   * @param null_precedence Indicates how null values are ordered with other values
+   * @param depth The depth of the column if part of a nested column @see preprocessed_table::depths
+   */
+  __device__ element_comparator(Nullate has_nulls,
+                                column_device_view lhs,
+                                column_device_view rhs,
+                                null_order null_precedence = null_order::BEFORE,
+                                int depth                  = 0)
+    : _lhs{lhs}, _rhs{rhs}, _nulls{has_nulls}, _null_precedence{null_precedence}, _depth{depth}
+  {
+  }
+
+  /**
+   * @brief Performs a relational comparison between the specified elements
+   *
+   * @param lhs_element_index The index of the first element
+   * @param rhs_element_index The index of the second element
+   * @return Indicates the relationship between the elements in the `lhs` and `rhs` columns, along
+   * with the depth at which a null value was encountered.
```

Comment:
Multi-lines doxygen throughout this PR is very difficult to read. Please align them like this:
```suggestion
   * @return Indicates the relationship between the elements in the `lhs` and `rhs` columns, along
   *         with the depth at which a null value was encountered.
```

Reply:
@ttnghia I don't think it's practical to expect people to hand format it like this. I suggest finding a way to create a clang-format rule that will do this automatically.